extractor/service.go

```diff
@@ -1,20 +1,240 @@
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/bblfsh/sdk/protocol"
+	"google.golang.org/grpc"
+
+	"gopkg.in/src-d/go-git.v4"                 // git.Open
+	"gopkg.in/src-d/go-git.v4/plumbing"        // Hash, Repository
+	"gopkg.in/src-d/go-git.v4/plumbing/object" // object.File
+	"gopkg.in/src-d/go-git.v4/storage"
+
+	"gopkg.in/src-d/core-retrieval.v0" // core_retrieval.RootTransactioner
+	"gopkg.in/src-d/core.v0"           // core.ModelRepositoryStore
+	"gopkg.in/src-d/core.v0/model"     // model.Repository, model.Reference
+	"gopkg.in/src-d/enry.v1"           // lang detection
 )
 
 type Service struct {
-	// TODO
+	bblfshClient protocol.ProtocolServiceClient
+}
+
+func NewService() *Service {
+	conn, err := grpc.Dial("0.0.0.0:9432", grpc.WithTimeout(time.Second*2), grpc.WithInsecure())
+	client := protocol.NewProtocolServiceClient(conn)
+	checkIfError(err)
+
+	return &Service{bblfshClient: client}
 }
 
 //proteus:generate
-func (*Service) GetRepositoryData(r *Request) (*RepositoryData, error) {
+func (s *Service) GetRepositoryData(r *Request) (*RepositoryData, error) {
 	// TODO
 	return nil, fmt.Errorf("NOT IMPLEMENTED YET")
 }
 
 //proteus:generate
-func (*Service) GetRepositoriesData() ([]*RepositoryData, error) {
-	return nil, fmt.Errorf("NOT IMPLEMENTED YET")
+func (s *Service) GetRepositoriesData() ([]*RepositoryData, error) {
+	n, err := core.ModelRepositoryStore().Count(model.NewRepositoryQuery().FindByStatus(model.Fetched))
+	checkIfError(err)
+	fmt.Printf("Iterating over %d 'fetched' repositories in DB:\n", n)
+
+	master := "refs/heads/master"
+	result := make([]*RepositoryData, n)
+
+	reposNum := 0
+	totalFilesNum := 0
+	for masterRefInit, repoID := range findAllFetchedReposWithRef(master) {
+		fmt.Printf("Repo: %v", repoID)
+		repo := &RepositoryData{
+			RepositoryID: repoID,
+			URL:          "", //TODO(bzz): add repo url!
+			Files:        make([]File, 100),
+		}
+
+		rootedTransactioner := core_retrieval.RootedTransactioner()
+		tx, err := rootedTransactioner.Begin(plumbing.Hash(masterRefInit))
+		if err != nil {
+			fmt.Printf("Failed to begin a tx for repo:%v, hash:%v. %v\n", repoID, masterRefInit, err)
+			continue
+		}
+
+		tree, err := gitOpenGetTree(tx.Storer(), repoID, masterRefInit, master)
+		if err != nil {
+			fmt.Printf("Failed to open&get tree for Git repo:%v, hash:%v. %v\n", repoID, masterRefInit, err)
+			_ = tx.Rollback()
+			continue
+		}
+
+		skpFilesNum := 0
+		sucFilesNum := 0
+		errFilesNum := 0
+		err = tree.Files().ForEach(func(f *object.File) error {
+			i := (skpFilesNum + sucFilesNum + errFilesNum) % 1000
+			batch := (skpFilesNum + sucFilesNum + errFilesNum) / 1000
+			if i == 0 && batch != 0 {
+				fmt.Printf("\t%d000 files...\n", batch)
+			}
+
+			// discard vendoring with enry
+			if enry.IsVendor(f.Name) || enry.IsDotFile(f.Name) ||
+				enry.IsDocumentation(f.Name) || enry.IsConfiguration(f.Name) {
+				skpFilesNum++
+				return nil
+			} //TODO(bzz): filter binaries like .apk and .jar
```

Comment:
Not really. This is about filtering out huge binaries before reading the content in memory, to speed things up. Enry `isBinary` works in a different, classifying array of bites